File: AndroidBasics/notes/src/main/java/pan/alexander/notes/di/RoomModule.java
package pan.alexander.notes.di;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.room.Room;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import pan.alexander.notes.data.database.AppDatabase;
import pan.alexander.notes.data.database.NotesDao;
import pan.alexander.notes.data.database.TrashDao;

@Module
public class RoomModule {
    private final static String APP_DATABASE_NAME = "app-database";

    private final AppDatabase db;

    public RoomModule(@NonNull Application appContext) {
        db = Room.databaseBuilder(appContext, AppDatabase.class, APP_DATABASE_NAME).build();
    }

    @Provides
    @Singleton
    AppDatabase providesRoomDatabase() {
        return db;
    }

    @Provides
    @Singleton
    NotesDao providesNotesDao() {
        return db.getNotesDao();
    }

    @Provides
    @Singleton
    TrashDao providesTrashDao() {
        return db.getTrashDao();
    }
}

File: TestWeatherApp/app/src/main/java/pan/alexander/testweatherapp/di/RoomModule.kt
package pan.alexander.testweatherapp.di

import android.content.Context
import androidx.room.Room
import org.koin.android.ext.koin.androidContext
import org.koin.dsl.module
import pan.alexander.testweatherapp.framework.database.AppDatabase

private const val APP_DATABASE_NAME = "main_database"

object RoomModule {

    fun provideAppDatabase() = module {
        single { provideDatabase(androidContext()) }
        single { provideCurrentWeatherDao(appDatabase = get()) }
    }

    private fun provideDatabase(context: Context) = Room
        .databaseBuilder(context, AppDatabase::class.java, APP_DATABASE_NAME)
        .fallbackToDestructiveMigration()
        .build()

    private fun provideCurrentWeatherDao(appDatabase: AppDatabase) = appDatabase.currentWeatherDao()
}
